src/Http/ResponseEmitter.php

```diff
@@ -0,0 +1,274 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.3.5
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ *
+ * Parts of this file are derived from Zend-Diactoros
+ *
+ * @copyright Copyright (c) 2015-2016 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-diactoros/blob/master/LICENSE.md New BSD License
+ */
+namespace Cake\Http;
+
+use Cake\Core\Configure;
+use Cake\Log\Log;
+use Psr\Http\Message\ResponseInterface;
+use Zend\Diactoros\RelativeStream;
+use Zend\Diactoros\Response\EmitterInterface;
+
+/**
+ * Emits a Response to the PHP Server API.
+ *
+ * This emitter offers a few changes from the emitters offered by
+ * diactoros:
+ *
+ * - It logs headers sent using CakePHP's logging tools.
+ * - Cookies are emitted using setcookie() to not conflict with ext/session
+ */
+class ResponseEmitter implements EmitterInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function emit(ResponseInterface $response, $maxBufferLength = 8192)
+    {
+        $file = $line = null;
+        if (headers_sent($file, $line)) {
+            $message = "Unable to emit headers. Headers sent in file=$file line=$line";
+            if (Configure::read('debug')) {
+                trigger_error($message, E_USER_WARNING);
+            } else {
+                Log::warn($message);
```

Comment:
Can we always (no matter what debug is set to) have the warning go to the Log (or does trigger_error do that anway)?